app/routes.py

```diff
@@ -1,2 +1,338 @@
+#put updates the entire record and patch is used when you want to modify a particular property of a record
+
+def mark_complete(task_id):
+
+    task = Task.query.get(task_id)
+
+    if task is None:
+        return jsonify(None), 404
+    
+    title_of_task =  f"Someone just completed the task {task.title}" 
+    
+    requests.post("https://slack.com/api/chat.postMessage", data={"channel": "task-notifications", "channelid": "C02K8MA17KR", "text": title_of_task}, headers={"authorization": os.environ.get("SLACK_API_TOKEN")})
+    #we have to use the os.environ.get to get the value of the variable becacuse this is an environment variable. We know it's an environment variable because we created it in env file - 
+    #variable whose value is set outside of the regular program files and refers to something that has to do with operating system.
+
+    task.completed_at = datetime.today()
+
+    db.session.commit()
+
+    return {
+        "task": {
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete": False if task.completed_at == None else True
+        }
+    }
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods= ["PATCH"])
+
+
+
+def mark_incomplete(task_id):
+
+    task = Task.query.get(task_id)
+    # taskvariable = None
+    if task is None:
+        return jsonify(None), 404
+
+    task.completed_at = None
+    
+
+    db.session.commit()
+
+    return {
+        "task": {
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete": False if task.completed_at == None else True
+        }
+    }
+
+
+# goals_bp = Blueprint("goals_bp", __name__, url_prefix=("/goals"))
+@goals_bp.route("", methods=["GET", "POST"])
+#any time we make a request to our api and call localhost5000, we are going to run the function.
+#what will run depends on if the client calls GET or POST
+def handle_goals(): #view function names should be unique
+    if request.method == "GET":
+        goals = Goal.query.all() #Task is a child of SQLA so we can use the query method, 
+        #query is method from sqla, paired with .all, retrieves all rows from task table in database, from there SQLA will turn all those rows into instances
+        #of the task class, and all method will return all instances of the Task class into a list, and that list is assigned to variable task
+        #Task.query.all.sort?? we can order the result the result of our selection
+        goals_response = []
+        for goal in goals: #<Task 2342> - returns object, if we want to see attributes we have to 
+            #access the attributes. So we break down objhects in that list in dictionary so we can see each instance's attributes
+            goals_response.append({
+                "id": goal.goal_id,
+                "title": goal.title
+            })
+        return jsonify(goals_response)
```

Comment:
List comprehension and a helper method in the goal class can be handy here:

```suggestion
        goals_response = [goal.to_dict() for goal in goals]
        return jsonify(goals_response)
```